File: android/app/src/test/java/com/hisab/app/domain/MoneyTest.kt
package com.hisab.app.domain

import org.junit.Assert.assertEquals
import org.junit.Test

class MoneyTest {
    // Kotlin enforces "never a float" for Money at compile time (Money(Long)
    // simply cannot accept a Double) — stronger than the runtime check needed
    // on the TypeScript side, where `number` doesn't distinguish int/float.
    // Same invariant, different enforcement mechanism per platform (see
    // docs/PHASE_GUIDE.md M2's note on shared invariants, not shared code).

    @Test
    fun `money accepts integers, including negative for reversals`() {
        assertEquals(100L, Money(100).minorUnits)
        assertEquals(-50L, Money(-50).minorUnits)
        assertEquals(0L, Money.ZERO.minorUnits)
    }

    @Test
    fun `plus and minus stay integer`() {
        assertEquals(Money(700), Money(500) + Money(200))
        assertEquals(Money(500), Money(700) - Money(200))
    }

    @Test
    fun `sum sums a list, empty list sums to zero`() {
        assertEquals(Money(600), listOf(Money(100), Money(200), Money(300)).sum())
        assertEquals(Money.ZERO, emptyList<Money>().sum())
    }
}

File: server/src/domain/money.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { money, addMoney, subtractMoney, sumMoney, ZERO_MONEY } from './money.js'

test('money rejects non-integer values (never a float)', () => {
  assert.throws(() => money(10.5), TypeError)
  assert.throws(() => money(0.1), TypeError)
})

test('money accepts integers, including negative (for reversals)', () => {
  assert.equal(money(100), 100)
  assert.equal(money(-50), -50)
  assert.equal(money(0), 0)
})

test('addMoney and subtractMoney stay integer', () => {
  assert.equal(addMoney(money(500), money(200)), 700)
  assert.equal(subtractMoney(money(700), money(200)), 500)
})

test('sumMoney sums a list, empty list sums to zero', () => {
  assert.equal(sumMoney([money(100), money(200), money(300)]), 600)
  assert.equal(sumMoney([]), ZERO_MONEY)
})
